subsys/bluetooth/host/classic/avdtp.c

```diff
@@ -729,15 +845,59 @@ static void avdtp_abort_handler(struct bt_avdtp *session,
 			}
 			LOG_DBG("abort err code:%d", error_code);
 			net_buf_add_u8(rsp_buf, error_code);
-		} else {
-			sep->state = AVDTP_IDLE;
 		}
 
 		err = bt_l2cap_chan_send(&session->br_chan.chan, rsp_buf);
 		if (err < 0) {
 			net_buf_unref(rsp_buf);
 			LOG_ERR("Error:L2CAP send fail - result = %d", err);
 			return;
+		} else {
+			if ((sep->state & (AVDTP_OPEN | AVDTP_STREAMING)) &&
+			    (sep->chan.state == BT_L2CAP_CONNECTED)) {
+				bt_avdtp_set_state(sep, AVDTP_ABORTING);
+			} else {
+				bt_avdtp_set_state(sep, AVDTP_IDLE);
+			}
+		}
+	} else {
+		struct bt_avdtp_req *req = session->req;
+
+		if (session->req == NULL) {
+			return;
+		}
+		k_work_cancel_delayable(&session->timeout_work);
+		if (msg_type == BT_AVDTP_ACCEPT) {
+			uint8_t pre_state = CTRL_REQ(req)->sep->state;
+
+			req->status = 0;
+			bt_avdtp_set_state(CTRL_REQ(req)->sep, AVDTP_ABORTING);
+			/* release stream */
+			if (pre_state & (AVDTP_OPEN | AVDTP_STREAMING)) {
+				avdtp_media_disconnect(CTRL_REQ(req)->sep);
+			}
+
+			/* For abort, make sure the state revert to IDLE state after
+			 * releasing l2cap channel.
+			 */
+			bt_avdtp_set_state(CTRL_REQ(req)->sep, AVDTP_IDLE);
+		} else if (msg_type == BT_AVDTP_REJECT) {
+			uint8_t acp_seid;
+
+			acp_seid = net_buf_pull_u8(buf);
+			if (acp_seid != CTRL_REQ(req)->acp_stream_ep_id) {
+				return;
+			}
+
+			req->status = net_buf_pull_u8(buf);
+		} else if (msg_type == BT_AVDTP_GEN_REJECT) {
+			req->status = BT_AVDTP_NOT_SUPPORTED_COMMAND;
+		}
+		AVDTP_LOCK();
```

Comment:
You should really fix this global lock stuff. First of all, a semaphore would be better than a mutex, and second of all it should be per-context rather than a global variable.

Reply:
change to semaphore and move it to the avdtp instance struct.